nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/DocumentSetRepositoryInit.java

```diff
@@ -23,23 +23,69 @@
 import org.nuxeo.ecm.core.api.CoreSession;
 import org.nuxeo.ecm.core.api.DocumentModel;
 import org.nuxeo.ecm.core.api.PathRef;
+import org.nuxeo.ecm.core.api.security.ACE;
+import org.nuxeo.ecm.core.api.security.ACL;
+import org.nuxeo.ecm.core.api.security.ACP;
+import org.nuxeo.ecm.core.api.security.impl.ACLImpl;
+import org.nuxeo.ecm.core.api.security.impl.ACPImpl;
 import org.nuxeo.ecm.core.test.DefaultRepositoryInit;
 
 /**
  * @since 10.2
  */
 public class DocumentSetRepositoryInit extends DefaultRepositoryInit {
 
-    private static final int SIZE = 10;
+    public static final String USERNAME = "bob";
+
+    public static final String ROOT = "/default-domain/workspaces/test";
+
+    public static final int SIZE = 3;
+
+    public static final int DOC_BY_LEVEL = SIZE * 2;
+
+    public static int created_non_proxy;
+
+    public static int created_proxy;
+
+    public static int created_total;
+
+    static {
+        // with SIZE = 3, 9 + 27 + 81 = 117 documents are created
+        for (int i = 1; i <= SIZE; i++) {
+            created_total += 3 * Math.pow(SIZE, i);
+        }
+        created_proxy = created_total / 3;
+        created_non_proxy = created_total / 3 * 2;
+    }
 
     @Override
     public void populate(CoreSession session) {
         super.populate(session);
-        DocumentModel test = session.getDocument(new PathRef("/default-domain/workspaces/test"));
-        for (int i = 0; i < SIZE; i++) {
-            DocumentModel doc = session.createDocumentModel(test.getPathAsString(), "doc" + i, "ComplexDoc");
-            doc.setProperty("dublincore", "modified", new Date());
-            session.createDocument(doc);
+        DocumentModel test = session.getDocument(new PathRef(ROOT));
+        ACP acp = new ACPImpl();
+        ACL acl = new ACLImpl();
+        acl.add(new ACE("Administrator", "Everything", true));
+        acl.add(new ACE(USERNAME, "WriteProperties", true));
+        acl.add(new ACE(USERNAME, "Read", true));
+        acp.addACL(acl);
+        test.setACP(acp, false);
+        createChildren(session, test, SIZE);
+    }
+
+    private void createChildren(CoreSession s, DocumentModel p, int depth) {
+        if (depth > 0) {
+            for (int i = 0; i < SIZE; i++) {
+                DocumentModel child = s.createDocumentModel(p.getPathAsString(), p.getName() + "doc" + i, "ComplexDoc");
+                child.setProperty("dublincore", "modified", new Date());
+                child = s.createDocument(child);
+                s.saveDocument(child);
```

Comment:
You don't need to `saveDocument` if you've just done a `createDocument` (here and further below)